src/perm/factored_transversal.rs

```diff
@@ -0,0 +1,173 @@
+use super::Permutation;
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::collections::VecDeque;
+
+use std::rc::Rc;
+///Represents a Factored Traversal/Schrier Vector of an elements orbit.
+/// Contains the base of this traversal, and a factored traversal of the orbit.
+#[derive(Debug)]
+pub struct FactoredTransversal {
+    // Base point of the orbit
+    base: usize,
+    // The factored traversal/Schrier vector of the orbit.
+    transversal: HashMap<usize, Rc<Permutation>>,
+}
+
+impl FactoredTransversal {
+    /// Given a set of generating elements and element, construct the factored traversal.
+    ///```
+    /// use stabchain::perm::factored_transversal::FactoredTransversal;
+    /// use stabchain::perm::Permutation;
+    /// let fc = FactoredTransversal::from_generators(1, vec![Permutation::from(vec![1, 0])]);
+    ///```
+    pub fn from_generators(base: usize, gens: Vec<Permutation>) -> Self {
+        let mut transversal = HashMap::new();
+        let id = Permutation::id();
+        let id_ref = Rc::new(id);
+        transversal.insert(base, id_ref);
+        // Orbit elements that have not been used yet.
+        let mut to_traverse = VecDeque::new();
+        to_traverse.push_back(base);
+        // While there are still elements of the orbit unused.
+        while !to_traverse.is_empty() {
+            //Take an unused element.
+            let delta = to_traverse.pop_front().unwrap();
+            for g in &gens {
+                let gamma = g.apply(delta);
+                // If the orbit doensn't contain this value, then add it to the factored transversal.
+                if let Entry::Vacant(v) = transversal.entry(gamma) {
+                    //Insert the inverse, to make calculating representatives easier
+                    v.insert(Rc::new(g.inv()));
+                    to_traverse.push_back(gamma);
```

Comment:
So, here I would use the Entry API rather than directly checking the Entry::Vacant, it is just a bit more idiomatic. I think get_or_insert_with will have the desired effect

Reply:
I'll look into that!